sage.py

```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
                 plt.ylabel('Micro attack stages')
-                # NOTE: Line thicknesses are on per host basis
-                tmax= max([x[4] for x in h_ep])
-                tmin = min([x[4] for x in h_ep])
-                for idx, ep in enumerate(h_ep):
-                    #print(idx, (ep[0], ep[1]), ep[2], ep[3][ep[0]:ep[1]+1])
-                    xax = list(range(ep[0], ep[1]+1))
-                    yax = [mcats.index(ep[2])]*len(xax)
-                    thickness  = ep[4]
-                    lsize = ((thickness - tmin)/ (tmax - tmin)) * (5-0.5) + 0.5 if (tmax - tmin) != 0.0 else 0.5
-                    #lsize = np.log(thickness) + 1 TODO: Either take log or normalize between [0.5 5]
-                    msize = (lsize*2)+1
+                # NOTE: Line thicknesses are on per-host basis
+                tmax = max([epi[4] for epi in host_episodes])
+                tmin = min([epi[4] for epi in host_episodes])
+                for idx, ep in enumerate(host_episodes):
+                    xax = list(np.arange(ep[0], ep[1]+1))
+                    yax = [mcats.index(ep[2])] * len(xax)
+                    thickness = ep[4]
+                    lsize = ((thickness - tmin) / (tmax - tmin)) * (5 - 0.5) + 0.5 if (tmax - tmin) != 0.0 else 0.5
+                    # lsize = np.log(thickness) + 1 TODO: Either take log or normalize between [0.5 5]
+                    msize = (lsize * 2) + 1
                     ax.plot(xax, yax, color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], linewidth=lsize)
                     ax.plot(ep[0], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0, markersize=msize, label=micro2macro[micro[ep[2]]])
                     ax.plot(ep[1], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0, markersize=msize)
-                    plt.yticks(range(len(mcats)), mcats_lab, rotation='0')
+                    plt.yticks(range(len(mcats)), [x.split('.')[1] for x in micro.values()], rotation=0)
                 legend_without_duplicate_labels(ax)
                 plt.grid(True, alpha=0.4)
                 
-                #plt.tight_layout()
-                #plt.savefig('Pres-Micro-attack-episodes-Team'+str(tid) +'-Connection'+ attacker[0]+'--'+attacker[1]+'.png')
+                # plt.tight_layout()
+                # plt.savefig('Pres-Micro-attack-episodes-Team'+str(tid) +'-Connection'+ attacker[0]+'--'+attacker[1]+'.png')
                 plt.show()
         
-        team_episodes.append(t_ep)
+        team_episodes_.append(team_host_episodes)
         team_times.append(_team_times)
-    return (team_episodes, team_times)
-            
+    return team_episodes_, team_times
+
+
 ## 17
 
 #### Host = [connections] instead of team level representation
 def host_episode_sequences(team_episodes):
     host_data = {}
-    #team_host_data= []
     print('# teams ', len(team_episodes))
     print('----- TEAMS -----')
```

Comment:
Teams are printed twice in the logs: during episode generation and during subsequence generation. Is it intentional?

Reply:
If you mean print(tid, sep=' ', end=' ', flush=True ) and print((tid+1), sep=' ', end=' ', flush=True), then I think the second one prints total number of subsequences, not teams. If that's true, we can rename the tid variable.

Reply:
Not really. The first time it is called in `aggregate_into_episodes`, after alerts are grouped per team. The second time is in `host_episode_sequences` method when iterating over episodes groups by team.

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5a07ddb7-5228-40d6-b9c8-fea028a9cd66)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/e06f04e2-911d-465d-8fd5-3587e8a408c0)